### PR TITLE
feat: add algorithm to find the number of paths in a graph using DFS

### PR DESCRIPTION
**Pull request by adi776borate**
#### Description of Change
- Added a new function to count the number of paths in a graph.
- Implemented a test case to validate the functionality of the new function.
- Improved documentation for better understanding of the algorithm.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: The new function counts the number of unique paths between two nodes in a graph. It can be particularly useful in graph-related problems where pathfinding is necessary.

### CODE REVIEW

File: graph/number_of_paths.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * @returns void
+ */
+static void test() {
+    // Test case 1: Simple directed graph with multiple paths
+    std::vector<std::vector<int>> graph1 = {
+        {0, 1, 0, 1, 0}, 
+        {0, 0, 1, 0, 1}, 
+        {0, 0, 0, 0, 1}, 
+        {0, 0, 1, 0, 0}, 
+        {0, 0, 0, 0, 0}
```
Comment (realstealthninja) [inline graph/number_of_paths.cpp:63]: add more tests featuring negative numbers as well or use unsigned integer as parameter

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Comment by adi776borate** [comment]
I've updated the algorithm and added more test cases as per the feedback received. Please review the latest changes. Thank you!

File: graph/number_of_paths.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+     * @brief Counts the number of paths from node `u` to node `v` in a directed graph
+     * using Depth First Search (DFS)
+     * 
+     * @param A adjacency matrix representing the graph (1: edge exists, 0: no edge)
+     * @param u the starting node
+     * @param v the destination node
+     * @param n the number of nodes in the graph
+     * @returns the number of paths from node `u` to node `v`
+     */
+    int count_paths(const std::vector<std::vector<unsigned int>>& A, int u, int v, int n) {
```
Comment (realstealthninja) [inline graph/number_of_paths.cpp:61]: since its impossible for number of paths to be negative use unsigned integers as the return type

File: graph/number_of_paths.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * @brief Helper function to perform DFS and count the number of paths from node `u` to node `v`
+     * @param A adjacency matrix representing the graph (1: edge exists, 0: no edge)
+     * @param u the starting node
+     * @param v the destination node
+     * @param n the number of nodes in the graph
+     * @param visited a vector to keep track of visited nodes in the current DFS path
+     * @returns the number of paths from node `u` to node `v`
+     */
+    int count_paths_dfs(const std::vector<std::vector<unsigned int>>& A, int u, int v, int n, std::vector<bool>& visited) {
```
Comment (realstealthninja) [inline graph/number_of_paths.cpp:33]: since its impossible for the number of paths from u to v to be negative use unsigned integer as the return type

Comment (realstealthninja) [inline graph/number_of_paths.cpp:33]:
```suggestion
    int count_paths_dfs(const std::vector<std::vector<unsigned int>>& A, int u, int v, int n, std::vector<bool>& visited) {
```
you've used `unsigned int` but thats a variable sized integer, Prefer fixed size integer
for example
```
std::uint32_t number // equivalent to `unsigned int number`

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: graph/number_of_paths.cpp (hunk 2 above)
Comment (adi776borate) [reply graph/number_of_paths.cpp]: I have made the changes as you requested; please review them. Thank you.

File: graph/number_of_paths.cpp (hunk 3 above)
Comment (adi776borate) [reply graph/number_of_paths.cpp]: I have made the changes as you requested; please review them. Thank you.

Comment (adi776borate) [reply graph/number_of_paths.cpp]: I have made the changes as you requested; please review them. Thank you.

File: graph/number_of_paths.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    std::vector<std::vector<std::uint32_t>> graph3 = {
+        {0, 1, 0, 0, 0}, 
+        {0, 0, 1, 1, 0}, 
+        {1, 0, 0, 0, 1}, 
+        {0, 0, 1, 0, 1}, 
+        {0, 0, 0, 0, 0}  
+    };
+    std::uint32_t n3 = 5, u3 = 0, v3 = 4;
+    assert(graph::count_paths(graph3, u3, v3, n3) == 3);  // There are 3 paths from node 0 to 4
+
```
Comment (realstealthninja) [inline graph/number_of_paths.cpp:113]: add a test case for empty array
std::vector<std::vector<std::uint32_t>> graph5 = {{}};

this should return 0 as there is no path

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: graph/number_of_paths.cpp (hunk 4 above)
Comment (adi776borate) [reply graph/number_of_paths.cpp:119]: I've made the latest changes as per your feedback. I hope this meets the requirements now. Looking forward to your review.

File: graph/number_of_paths.cpp [hunk 5]
```diff
@@ -0,0 +1,151 @@
+/**
```
[the remark this answers is not in the document]
Comment (adi776borate) [reply graph/number_of_paths.cpp:1]: Sure, it uses DFS to count the number of paths, but it is not just performing the traversal; naming it depth_first_search will create confusion and won't truly represent the purpose. If you think it must be renamed, I will be happy to do so.

[the remark this answers is not in the document]
Comment (adi776borate) [reply graph/number_of_paths.cpp:1]: I kindly request you to please add  "hactoberfest-accepted" label to this pull request, once you accept it. Thank You!

**Review by realstealthninja — APPROVED** [review]
Lgtm

**Review by realstealthninja — APPROVED** [review]
